### ISSUE TITLE
Update proxy agent config generation to TCP src/dst routing

### ISSUE THREAD

**Issue by ayj** [istio/manager#81]
lyft/envoy/pull/377 allows the tcp_proxy filter to pick the destination cluster based on a combination of L4 connection parameters (source/destination IP address/port). As the PR notes, this is a breaking change and old tcp_proxy configs will be rejected. The proxy agent needs to be updated to generate the newer tcp_proxy filter configuration.

This change must be synchronized with rebasing istio/proxy to pick up the latest version of lyft/envoy.

cc @qiwzhang, @kyessenov

**Comment by kyessenov** [comment]
Yes, I pulled this change from upstream here. Want to take a stab at this? I am a bit worried that we are all hacking on Envoy config generator right now, and will have many conflicts.

**Comment by ayj** [comment]
Sure. re: conflicts, we might need to coordinate PR if rebase conflicts become a problem.

**Comment by rshriram** [comment]
So I am working on the config gen as well. And one part of this is using the routing rules to generate tcp proxy blocks. 

However as part of this change, I am beginning to think that we need to reorganize the code to even more cleanly separate envoy config generation from the set of preprocessed inputs to the config generator. 

In addition, its unclear how one would go about in an optimized fashion, map the route match condition to a listener or virtual host, followed by the route blocks. A similar problem applies to the tcp proxy as well. In this case, it's even more arcane, as the current proto requires the user to specify IP port and subnet mask combo. 

I think we might have to revisit the proto spec and change it such that the user can specify the destination service as part of the match condition.

**Comment by ayj** [comment]
I believe the original issue is partially addressed by #84 which updates resource.go with the new envoy configuration definitions. I think the remaining piece there is to hook things up to the envoy config generation so the existing integration test passes with the latest envoy? @rshriram, were you going to address that in #84 or should I take that up after #84 goes in?

The next piece, which @rshriram is working on, is to hook these config generation up to the route spec which would be a separate issue/PR.

**Comment by rshriram** [comment]
So, the tests are passing finally. I think you can go ahead. May be keep an eye out for the stuff that @ijsnellf 's #87 is doing, when you add in code for the TCP blocks, so that its easier to wire in routing info in the selection process. However, it shouldn't be a blocker.

**Comment by ayj** [comment]
Envoy TCP config definitions have been added. Closing this issue as the TCP routing work is tracked by https://github.com/istio/manager/issues/46
